driver/src/main/com/mongodb/reactivestreams/client/internal/GridFSDownloadPublisherImpl.java

```diff
@@ -0,0 +1,214 @@
+        };
+
+        class GridFSDownloadStreamSubscriber implements Subscriber<Integer> {
+            private final ByteBuffer byteBuffer;
+
+            GridFSDownloadStreamSubscriber(final ByteBuffer byteBuffer) {
+                this.byteBuffer = byteBuffer;
+            }
+
+            @Override
+            public void onSubscribe(final Subscription s) {
+                s.request(1);
+            }
+
+            @Override
+            public void onNext(final Integer integer) {
+                synchronized (lock) {
+                    sizeRead += integer;
+                }
+            }
+
+            @Override
+            public void onError(final Throwable t) {
+                terminate();
+                outerSubscriber.onError(t);
+            }
+
+            @Override
+            public void onComplete() {
+                if (byteBuffer.remaining() > 0) {
+                    gridFSDownloadStream.read(byteBuffer).subscribe(new GridFSDownloadStreamSubscriber(byteBuffer));
+                } else {
+                    synchronized (lock) {
+                        requestedData = false;
+                        if (sizeRead == gridFSFile.getLength()) {
+                            isCompleted = true;
+                        }
+                    }
+                    byteBuffer.flip();
+                    outerSubscriber.onNext(byteBuffer);
+                    requestMoreOrComplete();
+                }
+            }
+        }
+
+        @Override
+        public void request(final long n) {
+            synchronized (lock) {
+                requested += n;
+            }
+            requestMoreOrComplete();
+        }
+
+        @Override
+        public void cancel() {
+            terminate();
+        }
+
+        private void requestMoreOrComplete() {
+            synchronized (lock) {
+                if (requested > 0 && !requestedData && !isTerminated && !isCompleted) {
+                    requestedData = true;
+                    if (gridFSFile == null) {
+                        getGridFSFile().subscribe(gridFSFileSubscriber);
+                    } else {
+                        requested--;
+                        int chunkSize = gridFSFile.getChunkSize();
+                        long remaining = gridFSFile.getLength() - sizeRead;
+
+                        if (remaining == 0) {
+                            isCompleted = true;
+                            requestedData = false;
+                            requestMoreOrComplete();
+                        } else {
+                            int byteBufferSize = Math.max(chunkSize, bufferSizeBytes);
+                            byteBufferSize =  Math.min(Long.valueOf(remaining).intValue(), byteBufferSize);
+                            int batchSize = Math.max(byteBufferSize / chunkSize, 1);
+                            ByteBuffer byteBuffer = ByteBuffer.allocate(byteBufferSize);
+                            if (batchSize != currentBatchSize) {
```

Comment:
Changing the batchsize has the side effect of throwing the cursor away. So only change the batchsize if the value changes.